### ISSUE TITLE
suppress exception logging during testing?

### ISSUE THREAD

**Issue by emiltin** [socketry/async#63]
In some RSpec tests, I expect an async task to timeout and raise Async::TimeoutError. It works, but the task exception prints to the terminal, which I would rather avoid in this case, since it interferes with the RSpec output

```
code/rsmp % be rspec spec/archive_spec.rb
..... 0.81s    error: Async::Task [oid=0x1518] [pid=52153] [2020-04-05 22:43:49 +0200]
               |   Async::TimeoutError: execution expired
               |   → /Users/emiltin/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/async-1.23.0/lib/async/task.rb:64 in `yield'
               |     /Users/emiltin/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/async-1.23.0/lib/async/condition.rb:38 in `wait'
               |     /Users/emiltin/Documents/code/rsmp/lib/rsmp/probe.rb:51 in `block in capture'
               |     /Users/emiltin/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/async-1.23.0/lib/async/reactor.rb:266 in `with_timeout'
               |     /Users/emiltin/.rbenv/versions/2.7.0/lib/ruby/2.7.0/forwardable.rb:235 in `with_timeout'
               |     /Users/emiltin/Documents/code/rsmp/lib/rsmp/probe.rb:50 in `capture'
               |     /Users/emiltin/Documents/code/rsmp/lib/rsmp/archive.rb:50 in `capture'
               |     /Users/emiltin/Documents/code/rsmp/spec/archive_spec.rb:122 in `block (4 levels) in <top (required)>'
               |     /Users/emiltin/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/async-1.23.0/lib/async/task.rb:255 in `block in make_fiber'
.

Finished in 0.18939 seconds (files took 0.7485 seconds to load)
6 examples, 0 failures
```

Is there a way to suppress the error output?

**Comment by ioquatix** [comment]
I've been burnt before from systems which suppress errors during testing.

If the timeout is an expected part of the code, you can catch it and deal with it (e.g. ignore it).

https://github.com/rsmp-nordic/rsmp/blob/c6958824b77380cc3c39f267c568e2a9ef3ee8ba/lib/rsmp/probe.rb#L49-L51

I think you should rescue timeout error.

The 2nd argument to with_timeout is the kind of exception to raise.

A timeout failing must raise an exception to abort the normal flow control. It's up to you to rescue it and behave accordingly.

Specifically:

- This is considered an error and you want an error to be displayed in production but not in testing, or,
- This is considered normal flow control, the timeout happened which is exceptional, so catch it and handle it (e.g. ignore) appropriately.

**Comment by ioquatix** [comment]
By the way, tasks where an exception propagates up, will print the exception, *unless* someone has called `task.wait` before the exception propagated up. So one option to surpress this error would be something like:

```ruby
task = Async do
  with_timeout { whatever }
end

expect{task.wait}.to raise_error(Async::TimeoutError)
```

Because you call task.wait, the error should not be printed out. However, I will point out there is a race condition between the code throwing an exception and the user calling wait. In the worst situation you will get an error printed out even though you eventually call `#wait`. I'm considering adding an option to task construction to suppress this behaviour.

**Comment by emiltin** [comment]
Thank you for your reply. Yes it's just the output I want to suppress, not the exception. So It's your second bullet: "This is considered normal flow control, the timeout happened which is exceptional, so catch it and handle it (e.g. ignore) appropriately". Just like suggest, I'm using:
```ruby
expect{task.wait}.to raise_error(Async::TimeoutError)
```

The test pass but the exception is printed. I will try your suggetion above using task.wait, but having race condition is not optimal, since the output might be printed sometimes.

**Comment by ioquatix** [comment]
```
expect{ do_something }.to raise_error(Async::TimeoutError)
```

If do_something is actually sleeping for a timeout, then there should be plenty of time for `task.wait` to suppress the error message. So, I actually think it's either a bug in your code, maybe you are not handling the exception correctly.

**Comment by emiltin** [comment]
Your suggestion still prints exception output, but maybe I'm missing something?

```ruby
task = Async do |task|
  task.with_timeout(1) { task.sleep 2 }
end
expect{task.wait}.to raise_error(Async::TimeoutError)
```
I think it's because the Async block is run synchronously and task.wait is therefore not called until after the timeout happens?

On the other hand, this code works (the test passes), and it does not prints the exception:
```ruby
Async do |task|
   timeout_task = task.async do
     task.with_timeout(1) { task.sleep 2 }
  end
  expect { timeout_task.wait }.to raise_error(Async::TimeoutError)
end
```

If the task.sleep is reduced to e.g. 0.5, the test fails as excepted:
```
Failures:

  1) RSMP::Archive timouts
     Failure/Error: expect { timeout_task.wait }.to raise_error(Async::TimeoutError)
       expected Async::TimeoutError but nothing was raised
     # ./spec/archive_spec.rb:153:in `block (3 levels) in <top (required)>'
```

**Comment by ioquatix** [comment]
Yes, you are right, it's because it's a top level block. There is no way to wait on it because it's synchronous.

One option is to use `async-rspec` and the reactor context.

The other option is to run it in a nested reactor.

I'll think about if I want to introduce a error suppression features. I'm not completely convinced it's a good idea.

**Comment by emiltin** [comment]
What do you think of the second option above (with the expect inside the Async block)? It seems to work.

**Comment by ioquatix** [comment]
I think the 2nd option is okay but it's still slightly racy. Here are some other options:

- Having an option to make a deferred task, which won't start until the next iteration of the run loop ensuring that your call to `wait` will hit the task before it starts executing, or
- Having an option to suppress error message output (e.g. injecting a custom logger which interleaves output with rspec if the spec fails).

**Comment by emiltin** [comment]
Thank you for the reply. Maybe a deferred task is most versatile. Would that work something like this?

```ruby
Async do |task|
   timeout_task = task.deferred do
     task.with_timeout(1) { task.sleep 2 }
  end
  expect { timeout_task.wait }.to raise_error(Async::TimeoutError)
end
```

**Comment by ioquatix** [comment]
You can now follow this approach to suppress logging:

https://github.com/socketry/async/blob/a2cabe6967a57ec94f97b7266a1f75c70fc36d0f/config/sus.rb#L5

**Comment by emiltin** [comment]
Thank you, I will try that.

**Comment by ioquatix** [comment]
When we merge inherited fiber variables https://bugs.ruby-lang.org/issues/19078 this won't be such a hack, and you'll be able to write something like `Console.logger.off!`.
